diff --git a/source/component/PasDoc_Parser.pas b/source/component/PasDoc_Parser.pas
index 02d5ee5d..b943a723 100644
--- a/source/component/PasDoc_Parser.pas
+++ b/source/component/PasDoc_Parser.pas
@@ -417,7 +417,7 @@ type
     procedure ParseInterfaceSection(const U: TPasUnit);
     procedure ParseImplementationSection(const U: TPasUnit);
     procedure ParseProperty(out p: TPasProperty);
-    procedure ParseType(const U: TPasUnit; IsGeneric: String);
+    procedure ParseType(const U: TPasUnit);
 
     { This assumes that you just read left parenthesis starting
       an enumerated type. It finishes parsing of TPasEnum,
@@ -1561,19 +1561,27 @@ begin
                 end;
               psType:
                 begin
-                  //In latest FPC version it is posible to define generic functions and procedures
+                  { "generic" directive may mean either
+                    - generic type or
+                    - generic function/procedure, so we need to look one token ahead.
+                    For generic type, call ParseType.
+                    for generic function/procedure, break -> thus ending
+                    ParseTVCSection, because we are no longer in var/const/type section. }
                   if t.IsStandardDirective(SD_GENERIC) then
                   begin
                     t2 := PeekNextToken;
                     if t2.IsKeyWord(KEY_FUNCTION) or t2.IsKeyWord(KEY_PROCEDURE) then
                       Break
                     else
-                      ParseType(U, t.Data)
+                    begin
+                      Scanner.UnGetToken(t); // put "generic" back, ParseType will consume it
+                      ParseType(U);
+                    end;
                   end
                   else
                   begin
-                    Scanner.UnGetToken(t);
-                    ParseType(U, '')
+                    Scanner.UnGetToken(t); // put initial identifier back, ParseType will consume it
+                    ParseType(U)
                   end;
                   AttributeIsPossible := True;
                 end;
@@ -2357,7 +2365,7 @@ begin
   end;
 end;
 
-procedure TParser.ParseType(const U: TPasUnit; IsGeneric: String);
+procedure TParser.ParseType(const U: TPasUnit);
 var
   RawDescriptionInfo: TRawDescriptionInfo;
   NormalType: TPasType;
@@ -2366,22 +2374,29 @@ var
   RoutineType: TPasRoutine;
   EnumType: TPasEnum;
   T, T2: TToken;
+  IsGeneric: boolean;
   CioType: TCIOType;
 begin
   { Read the type name, preceded by optional "generic" directive.
     Calculate TypeName, IsGeneric, TypeNameWithGeneric.
-    FPC requires "generic" directive, but Delphi doesn't,
+    FPC ObjFpc mode requires "generic" directive,
+    but Delphi (and FPC Delphi mode) don't,
     so it's just optional for us (serves for some checks later). }
   T := GetNextToken;
   try
-    if IsGeneric = '' then
-      TypeNameWithGeneric := ''
-    else
-      TypeNameWithGeneric := IsGeneric + ' ';
-
-    CheckToken(T, TOK_IDENTIFIER);
-    TypeName := T.Data;
+    TypeNameWithGeneric := '';
+    IsGeneric := T.IsStandardDirective(SD_GENERIC);
+    if IsGeneric then
+    begin
+      TypeNameWithGeneric := T.Data + ' ';
+      TypeName := GetAndCheckNextToken(TOK_IDENTIFIER);
+    end else
+    begin
+      CheckToken(T, TOK_IDENTIFIER);
+      TypeName := T.Data;
+    end;
     TypeNameWithGeneric := TypeNameWithGeneric + TypeName;
+
     FLastParsedTokenStreamAbsoluteFileName := T.StreamAbsoluteFileName;
     FLastParsedTokenLine := T.Line;
   finally FreeAndNil(T) end;
@@ -4101,7 +4116,8 @@ procedure TParser.ParseCio_Continue(const U: TPasUnit;
 
     { Read the type name, preceded by optional "generic" directive.
       Calculate TypeName, IsGeneric, TypeNameWithGeneric.
-      FPC requires "generic" directive, but Delphi doesn't,
+      FPC ObjFpc mode requires "generic" directive,
+      but Delphi (and FPC Delphi mode) don't,
       so it's just optional for us (serves for some checks later). }
     T := GetNextToken;
     try
@@ -4117,8 +4133,13 @@ procedure TParser.ParseCio_Continue(const U: TPasUnit;
         TypeName := T.Data;
       end;
       TypeNameWithGeneric := TypeNameWithGeneric + TypeName;
+
+      FLastParsedTokenStreamAbsoluteFileName := T.StreamAbsoluteFileName;
+      FLastParsedTokenLine := T.Line;
     finally FreeAndNil(T) end;
 
+    DoMessage(5, pmtInformation, 'Parsing type "%s"', [TypeName]);
+
     RawDescriptionInfo := GetLastComment;
     t := GetNextToken(LCollected);
     try
